// werift/lib/ice/src/stun/message.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
exports.parseMessage = parseMessage;
exports.paddingLength = paddingLength;
const crypto_1 = require("crypto");
const buffer_crc32_1 = __importDefault(require("buffer-crc32"));
const helper_1 = require("../helper");
const attributes_1 = require("./attributes");
const const_1 = require("./const");
function parseMessage(data, integrityKey) {
    if (data.length < const_1.HEADER_LENGTH) {
        return undefined;
    }
    const length = data.readUint16BE(2);
    if (data.length !== const_1.HEADER_LENGTH + length) {
        return undefined;
    }
    const messageType = data.readUint16BE(0);
    const transactionId = Buffer.from(data.slice(const_1.HEADER_LENGTH - 12, const_1.HEADER_LENGTH));
    const attributeRepository = new attributes_1.AttributeRepository();
    for (let pos = const_1.HEADER_LENGTH; pos <= data.length - 4;) {
        const attrType = data.readUInt16BE(pos);
        const attrLen = data.readUInt16BE(pos + 2);
        const payload = data.slice(pos + 4, pos + 4 + attrLen);
        const padLen = 4 * Math.floor((attrLen + 3) / 4) - attrLen;
        const attributesTypes = Object.keys(attributes_1.ATTRIBUTES_BY_TYPE);
        if (attributesTypes.includes(attrType.toString())) {
            const [, attrName, , attrUnpack] = attributes_1.ATTRIBUTES_BY_TYPE[attrType];
            if (attrUnpack.name === attributes_1.unpackXorAddress.name) {
                attributeRepository.setAttribute(attrName, attrUnpack(payload, transactionId));
            }
            else {
                attributeRepository.setAttribute(attrName, attrUnpack(payload));
            }
            if (attrName === "FINGERPRINT") {
                const fingerprint = messageFingerprint(data.slice(0, pos));
                const expect = attributeRepository.getAttributeValue("FINGERPRINT");
                if (expect !== fingerprint) {
                    return undefined;
                }
            }
            else if (attrName === "MESSAGE-INTEGRITY") {
                if (integrityKey) {
                    const integrity = messageIntegrity(data.slice(0, pos), integrityKey);
                    const expect = attributeRepository.getAttributeValue("MESSAGE-INTEGRITY");
                    if (!integrity.equals(expect)) {
                        return undefined;
                    }
                }
            }
        }
        pos += 4 + attrLen + padLen;
    }
    const attributes = attributeRepository.getAttributes();
    attributeRepository.clear();
    return new Message(messageType & 0x3eef, messageType & 0x0110, transactionId, attributes);
}
class Message extends attributes_1.AttributeRepository {
    constructor(messageMethod, messageClass, transactionId = (0, helper_1.randomTransactionId)(), attributes = []) {
        super(attributes);
        Object.defineProperty(this, "messageMethod", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: messageMethod
        });
        Object.defineProperty(this, "messageClass", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: messageClass
        });
        Object.defineProperty(this, "transactionId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: transactionId
        });
    }
    toJSON() {
        return this.json;
    }
    get json() {
        return {
            messageMethod: const_1.methods[this.messageMethod],
            messageClass: const_1.classes[this.messageClass],
            attributes: this.attributes,
        };
    }
    get transactionIdHex() {
        return this.transactionId.toString("hex");
    }
    get bytes() {
        let data = Buffer.from([]);
        for (const attrName of this.attributesKeys) {
            const attrValue = this.getAttributeValue(attrName);
            const [attrType, , attrPack] = attributes_1.ATTRIBUTES_BY_NAME[attrName];
            const v = attrPack.name === attributes_1.packXorAddress.name
                ? attrPack(attrValue, this.transactionId)
                : attrPack(attrValue);
            const attrLen = v.length;
            const padLen = 4 * Math.floor((attrLen + 3) / 4) - attrLen;
            const attrHeader = Buffer.alloc(4);
            attrHeader.writeUInt16BE(attrType, 0);
            attrHeader.writeUInt16BE(attrLen, 2);
            data = Buffer.concat([data, attrHeader, v, Buffer.alloc(padLen, 0)]);
        }
        const buf = Buffer.alloc(8);
        buf.writeUInt16BE(this.messageMethod | this.messageClass, 0);
        buf.writeUInt16BE(data.length, 2);
        buf.writeUInt32BE(const_1.COOKIE, 4);
        return Buffer.concat([buf, this.transactionId, data]);
    }
    addMessageIntegrity(key) {
        this.setAttribute("MESSAGE-INTEGRITY", this.messageIntegrity(key));
        return this;
    }
    messageIntegrity(key) {
        const checkData = setBodyLength(this.bytes, this.bytes.length - const_1.HEADER_LENGTH + const_1.INTEGRITY_LENGTH);
        return Buffer.from((0, crypto_1.createHmac)("sha1", key).update(checkData).digest("hex"), "hex");
    }
    addFingerprint() {
        this.setAttribute("FINGERPRINT", messageFingerprint(this.bytes));
    }
}
exports.Message = Message;
const setBodyLength = (data, length) => {
    const output = Buffer.alloc(data.length);
    data.copy(output, 0, 0, 2);
    output.writeUInt16BE(length, 2);
    data.copy(output, 4, 4);
    return output;
};
function messageFingerprint(data) {
    const checkData = setBodyLength(data, data.length - const_1.HEADER_LENGTH + const_1.FINGERPRINT_LENGTH);
    const crc32Buf = (0, buffer_crc32_1.default)(checkData);
    const xorBuf = Buffer.alloc(4);
    xorBuf.writeInt32BE(const_1.FINGERPRINT_XOR, 0);
    const fingerprint = (0, helper_1.bufferXor)(crc32Buf, xorBuf);
    return fingerprint.readUInt32BE(0);
}
function messageIntegrity(data, key) {
    const checkData = setBodyLength(data, data.length - const_1.HEADER_LENGTH + const_1.INTEGRITY_LENGTH);
    return Buffer.from((0, crypto_1.createHmac)("sha1", key).update(checkData).digest("hex"), "hex");
}
function paddingLength(length) {
    const rest = length % 4;
    if (rest === 0) {
        return 0;
    }
    else {
        return 4 - rest;
    }
}
