// werift/lib/webrtc/src/media/rtpTransceiver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Directions = exports.Sendrecv = exports.Recvonly = exports.Sendonly = exports.Inactive = exports.RTCRtpTransceiver = void 0;
const uuid = __importStar(require("uuid"));
const common_1 = require("../imports/common");
const const_1 = require("../const");
const stats_1 = require("./stats");
class RTCRtpTransceiver {
    set codecs(codecs) {
        this._codecs = codecs;
    }
    get codecs() {
        return this._codecs;
    }
    constructor(kind, dtlsTransport, receiver, sender, 
    /**RFC 8829 4.2.4.  direction the transceiver was initialized with */
    _direction) {
        Object.defineProperty(this, "kind", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: kind
        });
        Object.defineProperty(this, "receiver", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: receiver
        });
        Object.defineProperty(this, "sender", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: sender
        });
        Object.defineProperty(this, "_direction", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: _direction
        });
        Object.defineProperty(this, "id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: uuid.v4()
        });
        Object.defineProperty(this, "onTrack", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new common_1.Event()
        });
        Object.defineProperty(this, "mid", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "mLineIndex", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**should not be reused because it has been used for sending before. */
        Object.defineProperty(this, "usedForSender", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "_currentDirection", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "offerDirection", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_codecs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "headerExtensions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: {}
        });
        Object.defineProperty(this, "stopping", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "stopped", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        if (dtlsTransport) {
            this.setDtlsTransport(dtlsTransport);
        }
    }
    get dtlsTransport() {
        return this.receiver.dtlsTransport;
    }
    /**RFC 8829 4.2.4. setDirectionに渡された最後の値を示します */
    get direction() {
        return this._direction;
    }
    setDirection(direction) {
        this._direction = direction;
        if (const_1.SenderDirections.includes(this._currentDirection ?? "")) {
            this.usedForSender = true;
        }
    }
    /**RFC 8829 4.2.5. last negotiated direction */
    get currentDirection() {
        return this._currentDirection;
    }
    setCurrentDirection(direction) {
        this._currentDirection = direction;
    }
    setDtlsTransport(dtls) {
        this.receiver.setDtlsTransport(dtls);
        this.sender.setDtlsTransport(dtls);
    }
    get msid() {
        return `${this.sender.streamId} ${this.sender.trackId}`;
    }
    addTrack(track) {
        const res = this.receiver.addTrack(track);
        if (res) {
            this.onTrack.execute(track, this);
        }
    }
    // todo impl
    // https://www.w3.org/TR/webrtc/#methods-8
    stop() {
        if (this.stopping) {
            return;
        }
        // todo Stop sending and receiving with transceiver.
        this.stopping = true;
    }
    getPayloadType(mimeType) {
        return this.codecs.find((codec) => codec.mimeType.toLowerCase().includes(mimeType.toLowerCase()))?.payloadType;
    }
    getCodecStats() {
        const timestamp = (0, stats_1.getStatsTimestamp)();
        const stats = [];
        if (!this.dtlsTransport) {
            return stats;
        }
        const transportId = (0, stats_1.generateStatsId)("transport", this.dtlsTransport.id);
        // Add codec stats for each codec
        for (const codec of this.codecs) {
            const codecStats = {
                type: "codec",
                id: (0, stats_1.generateStatsId)("codec", codec.payloadType, transportId),
                timestamp,
                payloadType: codec.payloadType,
                transportId,
                mimeType: codec.mimeType,
                clockRate: codec.clockRate,
                channels: codec.channels,
                sdpFmtpLine: codec.parameters,
            };
            stats.push(codecStats);
        }
        return stats;
    }
}
exports.RTCRtpTransceiver = RTCRtpTransceiver;
exports.Inactive = "inactive";
exports.Sendonly = "sendonly";
exports.Recvonly = "recvonly";
exports.Sendrecv = "sendrecv";
exports.Directions = [exports.Inactive, exports.Sendonly, exports.Recvonly, exports.Sendrecv];
